android/app/src/main/java/fi/fmi/mobileweather/LargeWidgetProvider.java

```diff
@@ -0,0 +1,173 @@
+package fi.fmi.mobileweather;
+
+import static android.view.View.GONE;
+import static android.view.View.VISIBLE;
+
+import android.app.PendingIntent;
+import android.content.Intent;
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+import android.graphics.Color;
+import android.util.Log;
+import android.widget.RemoteViews;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.text.DateFormat;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Date;
+import java.util.Iterator;
+
+public class LargeWidgetProvider extends BaseWidgetProvider {
+    @Override
+    protected int getLayoutResourceId() {
+        return R.layout.large_widget_layout;
+    }
+
+    @Override
+    protected void onPostExecute(JSONObject json, JSONArray json2, RemoteViews main, SharedPreferencesHelper pref) {
+
+        Intent intent = new Intent(context, MainActivity.class);
+        PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_IMMUTABLE);
+
+        // Get settings
+        String background = pref.getString("background", "transparent");
+        Log.d("Download json", "Background: " + background);
+
+        main = new RemoteViews(context.getPackageName(), getLayoutResourceId());
+
+        // Show normal view
+        main.setInt(R.id.normalLayout, "setVisibility", VISIBLE);
+        // Hide error view
+        main.setInt(R.id.errorLayout, "setVisibility", GONE);
+
+        main.setOnClickPendingIntent(R.id.mainLinearLayout, pendingIntent);
+
+        json = useNewOrStoredJsonObject(json, pref);
+        if (json == null) return;
+
+        if (background.equals("dark")) {
```

Comment:
This is pretty long method so you could create helper method to return colors and make this method shorter. Maybe also other widgets would benefit from that.

Reply:
Done.

Quite much code refactoring done now.
But the code should be easier to understand now.